# MOCHAMMAD EDO PRAMUDYA
## 23031554070

Buatlah eksperimen memakai data Parkinson (dimensi : 22, class: PD atau Healthy). Gunakan 100 sampel pertama sebagai training set dan sisanya (95 sampel) sebagai testing set. 

In [1]:
import pandas as pd

file_path = 'parkinson.csv'
data = pd.read_csv(file_path)

data

a        b        c        d        e        f        g        h  \
0    119.992  157.302   74.997  0.00784  0.00007  0.00370  0.00554  0.01109   
1    122.400  148.650  113.819  0.00968  0.00008  0.00465  0.00696  0.01394   
2    116.682  131.111  111.555  0.01050  0.00009  0.00544  0.00781  0.01633   
3    116.676  137.871  111.366  0.00997  0.00009  0.00502  0.00698  0.01505   
4    116.014  141.781  110.655  0.01284  0.00011  0.00655  0.00908  0.01966   
..       ...      ...      ...      ...      ...      ...      ...      ...   
190  174.188  230.978   94.261  0.00459  0.00003  0.00263  0.00259  0.00790   
191  209.516  253.017   89.488  0.00564  0.00003  0.00331  0.00292  0.00994   
192  174.688  240.005   74.287  0.01360  0.00008  0.00624  0.00564  0.01873   
193  198.764  396.961   74.904  0.00740  0.00004  0.00370  0.00390  0.01109   
194  214.289  260.277   77.973  0.00567  0.00003  0.00295  0.00317  0.00885   

           i      j  ...        n        o       p         q         r  \
0    0.04374  0.426  ...  0.06545  0.02211  21.033  0.414783  0.815285   
1    0.06134  0.626  ...  0.09403  0.01929  19.085  0.458359  0.819521   
2    0.05233  0.482  ...  0.08270  0.01309  20.651  0.429895  0.825288   
3    0.05492  0.517  ...  0.08771  0.01353  20.644  0.434969  0.819235   
4    0.06425  0.584  ...  0.10470  0.01767  19.649  0.417356  0.823484   
..       ...    ...  ...      ...      ...     ...       ...       ...   
190  0.04087  0.405  ...  0.07008  0.02764  19.517  0.448439  0.657899   
191  0.02751  0.263  ...  0.04812  0.01810  19.147  0.431674  0.683244   
192  0.02308  0.256  ...  0.03804  0.10715  17.883  0.407567  0.655683   
193  0.02296  0.241  ...  0.03794  0.07223  19.020  0.451221  0.643956   
194  0.01884  0.190  ...  0.03078  0.04398  21.209  0.462803  0.664357   

            s         t         u         v  Class  
0   -4.813031  0.266482  2.301442  0.284654     PD  
1   -4.075192  0.335590  2.486855  0.368674     PD  
2   -4.443179  0.311173  2.342259  0.332634     PD  
3   -4.117501  0.334147  2.405554  0.368975     PD  
4   -3.747787  0.234513  2.332180  0.410335     PD  
..        ...       ...       ...       ...    ...  
190 -6.538586  0.121952  2.657476  0.133050     PD  
191 -6.195325  0.129303  2.784312  0.168895     PD  
192 -6.787197  0.158453  2.679772  0.131728     PD  
193 -6.744577  0.207454  2.138608  0.123306     PD  
194 -5.724056  0.190667  2.555477  0.148569     PD  

[195 rows x 23 columns]

In [2]:
data.head(), data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 23 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   a       195 non-null    float64
 1   b       195 non-null    float64
 2   c       195 non-null    float64
 3   d       195 non-null    float64
 4   e       195 non-null    float64
 5   f       195 non-null    float64
 6   g       195 non-null    float64
 7   h       195 non-null    float64
 8   i       195 non-null    float64
 9   j       195 non-null    float64
 10  k       195 non-null    float64
 11  l       195 non-null    float64
 12  m       195 non-null    float64
 13  n       195 non-null    float64
 14  o       195 non-null    float64
 15  p       195 non-null    float64
 16  q       195 non-null    float64
 17  r       195 non-null    float64
 18  s       195 non-null    float64
 19  t       195 non-null    float64
 20  u       195 non-null    float64
 21  v       195 non-null    float64
 22  Cl

(         a        b        c        d        e        f        g        h  \
 0  119.992  157.302   74.997  0.00784  0.00007  0.00370  0.00554  0.01109   
 1  122.400  148.650  113.819  0.00968  0.00008  0.00465  0.00696  0.01394   
 2  116.682  131.111  111.555  0.01050  0.00009  0.00544  0.00781  0.01633   
 3  116.676  137.871  111.366  0.00997  0.00009  0.00502  0.00698  0.01505   
 4  116.014  141.781  110.655  0.01284  0.00011  0.00655  0.00908  0.01966   
 
          i      j  ...        n        o       p         q         r  \
 0  0.04374  0.426  ...  0.06545  0.02211  21.033  0.414783  0.815285   
 1  0.06134  0.626  ...  0.09403  0.01929  19.085  0.458359  0.819521   
 2  0.05233  0.482  ...  0.08270  0.01309  20.651  0.429895  0.825288   
 3  0.05492  0.517  ...  0.08771  0.01353  20.644  0.434969  0.819235   
 4  0.06425  0.584  ...  0.10470  0.01767  19.649  0.417356  0.823484   
 
           s         t         u         v  Class  
 0 -4.813031  0.266482  2.301442  0.28

## KNN

In [3]:
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report

X = data.drop(columns=['Class'])
y = data['Class']

X_train, X_test = X[:100], X[100:]
y_train, y_test = y[:100], y[100:]

knn = KNeighborsClassifier(n_neighbors=5)

knn.fit(X_train, y_train)

y_pred = knn.predict(X_test)
report = classification_report(y_test, y_pred, target_names=['Healthy', 'PD'])
accuracy = accuracy_score(y_test, y_pred)
print(report)
print(f'Akurasi KNN: {accuracy}')


              precision    recall  f1-score   support

     Healthy       0.85      0.63      0.73        71
          PD       0.38      0.67      0.48        24

    accuracy                           0.64        95
   macro avg       0.62      0.65      0.61        95
weighted avg       0.73      0.64      0.66        95

Akurasi KNN: 0.6421052631578947


Model KNN mencapai akurasi sekitar 64.2%, yang berarti model ini berhasil mengklasifikasikan sekitar 64% dari data uji dengan benar. Namun, akurasi ini masih dapat dikatakan rendah untuk aplikasi medis yang umumnya memerlukan akurasi yang lebih tinggi.

## NAIVE BAYES

In [5]:
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, classification_report

nb = GaussianNB()
nb.fit(X_train, y_train)
y_pred_nb = nb.predict(X_test)
accuracy_nb = accuracy_score(y_test, y_pred_nb)
report_nb = classification_report(y_test, y_pred_nb, target_names=['Healthy', 'PD'])
print(report_nb)
print(f'Akurasi Naive Bayes: {accuracy_nb}')


              precision    recall  f1-score   support

     Healthy       0.85      0.39      0.54        71
          PD       0.31      0.79      0.44        24

    accuracy                           0.49        95
   macro avg       0.58      0.59      0.49        95
weighted avg       0.71      0.49      0.51        95

Akurasi Naive Bayes: 0.49473684210526314


Naive Bayes hanya mencapai akurasi sekitar 49.5%, yang lebih rendah dibandingkan dengan KNN (64.2%). Ini menunjukkan bahwa Naive Bayes kurang mampu mengidentifikasi pola pada data ini dibandingkan KNN. Dengan akurasi di bawah 50%.

## DECISSON TREE

In [6]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, classification_report

dt = DecisionTreeClassifier(random_state=0)
dt.fit(X_train, y_train)
y_pred_dt = dt.predict(X_test)
accuracy_dt = accuracy_score(y_test, y_pred_dt)
report_dt = classification_report(y_test, y_pred_dt, target_names=['Healthy', 'PD'])
print(report_dt)
print(f'Akurasi Decision Tree: {accuracy_dt}')


              precision    recall  f1-score   support

     Healthy       0.80      0.23      0.35        71
          PD       0.27      0.83      0.40        24

    accuracy                           0.38        95
   macro avg       0.53      0.53      0.38        95
weighted avg       0.67      0.38      0.36        95

Akurasi Decision Tree: 0.37894736842105264


Dengan akurasi 37.9%, Decision Tree menunjukkan performa yang lebih rendah dibandingkan KNN (64.2%) dan Naive Bayes (49.5%). Akurasi ini menunjukkan bahwa Decision Tree kurang mampu mengklasifikasikan data Parkinson dengan benar.

## RANDOM FORREST

In [7]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report

rf = RandomForestClassifier(random_state=42)
rf.fit(X_train, y_train)
y_pred_rf = rf.predict(X_test)
accuracy_rf = accuracy_score(y_test, y_pred_rf)
report_rf = classification_report(y_test, y_pred_rf, target_names=['Healthy', 'PD'])
print(report_rf)
print(f'Akurasi Random Forest: {accuracy_rf}')


              precision    recall  f1-score   support

     Healthy       0.89      0.55      0.68        71
          PD       0.37      0.79      0.51        24

    accuracy                           0.61        95
   macro avg       0.63      0.67      0.59        95
weighted avg       0.76      0.61      0.63        95

Akurasi Random Forest: 0.6105263157894737




Dari hasil eksperimen dengan beberapa algoritma machine learning untuk mendeteksi Parkinson menggunakan data yang diberikan, diperoleh hasil akurasi sebagai berikut:

1. **Random Forest**: 61.05%
2. **K-Nearest Neighbors (KNN)**: 64.21%
3. **Naive Bayes**: 49.47%
4. **Decision Tree**: 37.89%

### Kesimpulan

- **KNN** menghasilkan akurasi tertinggi (64.21%) dibandingkan algoritma lain. Hal ini menunjukkan bahwa metode berbasis tetangga terdekat cukup efektif dalam mengenali pola data pasien Parkinson dan membedakan antara pasien yang sehat dan yang memiliki Parkinson.
- **Random Forest** menunjukkan performa yang baik dengan akurasi 61.05%, menandakan bahwa ensemble learning juga memberikan klasifikasi yang cukup kuat. Namun, performanya masih di bawah KNN dalam eksperimen ini.
- **Naive Bayes** dan **Decision Tree** memiliki akurasi yang lebih rendah, menunjukkan bahwa pendekatan berbasis probabilistik sederhana dan pohon keputusan tunggal mungkin tidak cukup kuat dalam menangani kompleksitas dataset ini.
- **Imbalance** dalam akurasi deteksi antara kelas "Healthy" dan "PD" pada beberapa algoritma (khususnya Random Forest) menunjukkan bahwa model dapat lebih sulit dalam mendeteksi kelas "PD". Ini bisa menjadi indikasi bahwa dataset memerlukan lebih banyak penyeimbangan kelas atau fitur tambahan yang lebih jelas memisahkan kedua kelas.

Secara keseluruhan, KNN adalah model terbaik dalam eksperimen ini, tetapi perlu dipertimbangkan lebih lanjut metode lain atau tuning hyperparameter untuk meningkatkan akurasi dan konsistensi dalam mendeteksi penyakit Parkinson.